### PR TITLE
Benchmarking/bfs directopt

### PR DESCRIPTION
**Pull request by nicelhc13**
Implement bfs-direction optimization.
This branch is diverged from the benchmarking/perf-analysis which is being evaluated.

### CODE REVIEW

**Review by amberhassaan — CHANGES_REQUESTED** [review]
Hi Hochan, I'll summarize my comments here:
1. Please measure the time spent in BitsetToWl and WlToBitset. I'd like to compare these times to  the total BFS time. If this time is significant, I have some ideas to improve it. But before we go down that path, it's necessary to measure first. 
2. Please point me to the implementation of katana::CreateTransposeGraph(pg)
3. Let's make the parameters alpha and beta configurable from commandline or make them environment variables.

**Comment by nicelhc13** [comment]
> Hi Hochan, I'll summarize my comments here:
> 
> 1. Please measure the time spent in BitsetToWl and WlToBitset. I'd like to compare these times to  the total BFS time. If this time is significant, I have some ideas to improve it. But before we go down that path, it's necessary to measure first.
> 2. Please point me to the implementation of katana::CreateTransposeGraph(pg)
> 3. Let's make the parameters alpha and beta configurable from commandline or make them environment variables.

Hi Amber,
1. I confirmed that these took less than 1% of the total computation time. 
2. This is the link for [CreateTransposeGraph()](https://github.com/KatanaGraph/katana/blob/e1f6e2b8b84dc2599e8dc1bf04eb47e99b8750a8/libgalois/src/PropertyGraph.cpp#L1030).
3. Sure, I will work on this soon.

If you don't mind, could you please check if my implementation is buggy or has correct logic?
Thank you for your evaluation.
Could you

File: lonestar/analytics/cpu/bfs/bfs_cli.cpp [hunk 1]
```diff
@@ -69,7 +69,10 @@ static cll::opt<BfsPlan::Algorithm> algo(
             BfsPlan::kAsynchronousTile, "AsyncTile", "Asynchronous tiled"),
         clEnumValN(BfsPlan::kAsynchronous, "Async", "Asynchronous"),
         clEnumValN(BfsPlan::kSynchronousTile, "SyncTile", "Synchronous tiled"),
-        clEnumValN(BfsPlan::kSynchronous, "Sync", "Synchronous")),
+        clEnumValN(BfsPlan::kSynchronous, "Sync", "Synchronous"),
+        clEnumValN(
+            BfsPlan::kSynchronousDirectOpt, "SyncDO",
+            "Synchronous direction optimization")),
     cll::init(BfsPlan::kSynchronousTile));
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/bfs/bfs_cli.cpp]: Change the default to direction optimization. That way CI tests will be triggered for your new code.

File: libgalois/src/analytics/bfs/bfs.cpp [hunk 2]
```diff
@@ -115,6 +116,30 @@ struct OneTilePushWrap {
   }
 };
 
+template <typename WL>
+void
+WlToBitset(WL& wl, katana::DynamicBitset& bitset) {
+  katana::do_all(
+      katana::iterate(wl), [&](const Graph::Node& src) { bitset.set(src); },
+      katana::steal(), katana::chunk_size<kChunkSize>(),
```
Comment (roshandathathri) [inline libgalois/src/analytics/bfs/bfs.cpp]: Drop steal.

File: libgalois/src/analytics/bfs/bfs.cpp [hunk 3]
```diff
@@ -115,6 +116,30 @@ struct OneTilePushWrap {
...
+BitsetToWl(const Graph& graph, const katana::DynamicBitset& bitset, WL& wl) {
+  wl.clear();
+  katana::do_all(
+      katana::iterate(graph),
+      [&](const Graph::Node& src) {
+        if (bitset.test(src)) {
+          wl.push(src);
+        }
+      },
+      katana::steal(), katana::chunk_size<kChunkSize>(),
```
Comment (roshandathathri) [inline libgalois/src/analytics/bfs/bfs.cpp]: drop steal

File: libgalois/src/analytics/bfs/bfs.cpp [hunk 4]
```diff
@@ -246,9 +271,127 @@ SynchronousAlgo(
...
+
+  katana::GAccumulator<uint64_t> writes_pull, writes_push;
+  writes_pull.reset();
+  writes_push.reset();
+
+  while (!next_frontier->empty()) {
+    std::swap(frontier, next_frontier);
+    next_frontier->clear();
+    // TODO(lhc) need to do parameter if (scout_count > edges_to_check / alpha)
+    if (scout_count > edges_to_check / 15) {
```
Comment (roshandathathri) [inline libgalois/src/analytics/bfs/bfs.cpp]: Please don't use constants like this. Have preprocessor defines at least.

File: libgalois/src/analytics/bfs/bfs.cpp [hunk 5]
```diff
@@ -246,9 +271,127 @@ SynchronousAlgo(
...
+        ++next_level;
+        old_num_work_items = work_items.reduce();
+        work_items.reset();
+
+        loop(
+            katana::iterate(*transpose_graph_topo),
+            [&](const typename Graph::Node& dst) {
+              auto& ddata = graph->GetData<BfsNodeDistance>(dst);
+              if (ddata == BfsImplementation::kDistanceInfinity) {
+                for (auto e : graph->edges(dst)) {
```
Comment (roshandathathri) [inline libgalois/src/analytics/bfs/bfs.cpp]: Shouldn't this transpose_graph_topo?

Rename transpose_graph_topo to transpose_graph everywhere.

**Review by roshandathathri — APPROVED** [review]
LGTM

File: libgalois/src/analytics/bfs/bfs.cpp (hunk 5 above)
Comment (nicelhc13) [reply libgalois/src/analytics/bfs/bfs.cpp]: No. `Graph` is TypedPropertyGraph, and IIUC, this is additional wrapper objects to handle properties. It does not have transposing or making symmetric. So what I did is to just change topology. `PropertyGraph` has transposing function. I just called this, transposed the topology, still node data are on `TypedPropertyGraph`, and so, access only those data. Therefore, I could not change any of them.

Comment (roshandathathri) [reply libgalois/src/analytics/bfs/bfs.cpp]: I understood that but it doesn't answer my question. You should be accessing in_edges here, which is edges of transposed_graph.

Don't use "topo". PropertyGraph has properties too.

Comment (roshandathathri) [reply libgalois/src/analytics/bfs/bfs.cpp]: The additional wrapper in TypedPropertyGraph is to give a typed view of those properties.

File: libgalois/src/PropertyGraph.cpp [hunk 6]
```diff
@@ -1027,23 +1027,23 @@ katana::CreateSymmetricGraph(katana::PropertyGraph* pg) {
...
   auto transpose = std::make_unique<katana::PropertyGraph>();
   if (topology.num_nodes() == 0) {
     return std::unique_ptr<PropertyGraph>(std::move(transpose));
   }
 
   // New transpose graph topology
-  auto out_indices = std::make_unique<katana::LargeArray<uint64_t>>();
-  auto out_dests = std::make_unique<katana::LargeArray<uint32_t>>();
+  katana::LargeArray<uint64_t> out_indices;
+  katana::LargeArray<uint32_t> out_dests;
```
Comment (roshandathathri) [inline libgalois/src/PropertyGraph.cpp]: This should be moved out. Into the caller.

Move out the allocation and pass the `GraphTopology` as a parameter. This function should only set the values in it.

Comment (nicelhc13) [reply libgalois/src/PropertyGraph.cpp]: Yes that's what I heard during the meeting. Yes I will do.

File: libgalois/include/katana/PropertyGraph.h [hunk 7]
```diff
@@ -680,11 +680,21 @@ CreateSymmetricGraph(PropertyGraph* pg);
...
-/// \param pg The original property graph
+/// \param topology The original property graph topology
+/// \param out_indices The out edge index array for the transposed graph
+/// \param out_dests The out edge destination array for the transposed graph
 /// \return The new transposed property graph by reversing the edges
+// TODO(lhc): hack for bfs-direct-opt
+KATANA_EXPORT Result<std::unique_ptr<PropertyGraph>>
+CreateTransposeGraphTopology(
+    const GraphTopology& topology, LargeArray<uint64_t>& out_indices,
+    LargeArray<uint32_t>& out_dests);
```
Comment (roshandathathri) [inline libgalois/include/katana/PropertyGraph.h]: According to our coding guidelines, `const Type&` for read-only parameters and `Type*` for read-write parameters.

File: libgalois/include/katana/PropertyGraph.h [hunk 8]
```diff
@@ -680,11 +680,21 @@ CreateSymmetricGraph(PropertyGraph* pg);
...
+
 // TODO(gill): Add tranposed edge properties as well.
+/*
 KATANA_EXPORT Result<std::unique_ptr<katana::PropertyGraph>>
-CreateTransposeGraph(PropertyGraph* pg);
+CreateTransposeGraphTopology(PropertyGraph* pg);
```
Comment (roshandathathri) [inline libgalois/include/katana/PropertyGraph.h]: Drop commented code.

File: lonestar/analytics/cpu/bfs/bfs_cli.cpp [hunk 9]
```diff
@@ -69,8 +69,19 @@ static cll::opt<BfsPlan::Algorithm> algo(
...
+            BfsPlan::kSynchronousDirectOpt, "SyncDO",
+            "Synchronous direction optimization")),
+    cll::init(BfsPlan::kSynchronousDirectOpt));
+
+static cll::opt<unsigned int> alpha(
+    "alpha", cll::desc("Alpha for direct optimization (default value: 15)"),
+    cll::init(15));
+
+static cll::opt<unsigned int> beta(
+    "beta", cll::desc("Beta for direct optimization (default value: 18)"),
```
Comment (roshandathathri) [inline lonestar/analytics/cpu/bfs/bfs_cli.cpp]: direction optimization in both places

**Review by roshandathathri — APPROVED** [review]
LGTM

File: lonestar/analytics/cpu/bfs/bfs_cli.cpp (hunk 9 above)
Comment (nicelhc13) [reply lonestar/analytics/cpu/bfs/bfs_cli.cpp]: OOops

**Review by amberhassaan — APPROVED** [review]
I'm good.

**Pull request merged**